### PR TITLE
[RFC] Improve complete() flicker

### PR DESCRIPTION
**Pull request by Shougo**
It improves the `complete()` flicker.

More testers with slow terminals are needed.
You can test the patch with deoplete.

### CODE REVIEW

**Comment by Shougo** [comment]
I will implement it in Vim after it is reviewed.

**Comment by Shougo** [comment]
Fixed the clint errors.

File: src/nvim/edit.c [hunk 1]
```diff
@@ -184,6 +184,9 @@ static expand_T compl_xp;
 
 static int compl_opt_refresh_always = FALSE;
 
+static bool compl_show_pum = true;          // FALSE: show popup menu
+                                            //  TRUE: don't show popup menu
```
Comment (cacplate) [inline src/nvim/edit.c]: why not `false` and `true`?

File: src/nvim/edit.c [hunk 2]
```diff
@@ -8508,3 +8512,22 @@ static char_u *do_insert_char_pre(int c)
 
   return res;
 }
+
+static void show_pum(int save_w_wrow)
+{
+  int n;
```
Comment (cacplate) [inline src/nvim/edit.c]: It is better to initialize at the declaration from the style guide.

**Comment by Shougo** [comment]
@cacplate Thank you!   Fixed.

File: src/nvim/edit.c [hunk 3]
```diff
@@ -2321,9 +2324,12 @@ static int ins_compl_make_cyclic(void)
  */
 void set_completion(colnr_T startcol, list_T *list)
 {
-  /* If already doing completions stop it. */
-  if (ctrl_x_mode != 0)
+  int save_w_wrow;
```
Comment (cacplate) [inline src/nvim/edit.c]: I should have seen this one too, sorry.

Comment (Shougo) [inline src/nvim/edit.c]: Fixed.

**Comment by Shougo** [comment]
@justinmk Can you merge it?

**Comment by Shougo** [comment]
I think it reduces the complete flicker.

https://github.com/zchee/deoplete-jedi/issues/6

**Comment by alexlafroscia** [comment]
I tested this out against `deoplete-jedi` (issue linked by @Shougo above) and can confirm that the flickering is fixed now.

**Comment by nhooyr** [comment]
Works fine for me, no issues so far.

File: src/nvim/edit.c [hunk 4]
```diff
@@ -184,6 +184,9 @@ static expand_T compl_xp;
 
 static int compl_opt_refresh_always = FALSE;
 
+static bool compl_show_pum = true;          // false: show popup menu
+                                            //  true: don't show popup menu
```
Comment (oni-link) [inline src/nvim/edit.c]: Is the meaning of `true` and `false` swapped here?

Comment (Shougo) [inline src/nvim/edit.c]: Oh, it was mistake.  Thanks.

**Comment by landaire** [comment]
Works great. I went from the issues shown in https://github.com/Shougo/deoplete.nvim/issues/170 to none.

**Comment by Shougo** [comment]
Fixed for the commend.  And rebased.

**Comment by Shougo** [comment]
> Edit: I'm not sure if this is the cause, but I can only seem to select the first or last option from the list now. Once I hit CTL-N or CTL-P to go to the first/last option, it automatically gets inserted and pressing the key a second time starts to complete the next word instead of moving to the next completion option.

I have tested, but I cannot reproduce the problem.
Please create minimal init.vim and reproduce ways.

**Comment by alexlafroscia** [comment]
@Shougo after investigating more, I think what I was seeing is actually a bug in `deoplete-jedi` specifically.  This patch is fixing the flickering error really well.  I made an issue describing that other issue here: https://github.com/zchee/deoplete-jedi/issues/7

**Comment by Shougo** [comment]
@justinmk Can you merge it?  It fixes deoplete flicker problem.
Some testers have tested and works well.

**Comment by fmoralesc** [comment]
LGTM

**Comment by justinmk** [comment]
This deserves a test. It should be possible to simulate the "interruption". This also ensures the codepath is checked by coverity.

**Comment by Shougo** [comment]
OK.  I will add the tests.

>  It should be possible to simulate the "interruption".

It cannot be interrupt.  It is just the redraw problem.

**Comment by Shougo** [comment]
Rebased.   I have checked the tests, but `complete()` tests are already added.
No need to add tests.

**Comment by justinmk** [comment]
`pum_wanted()` already exists, and now this introduces a new `compl_show_pum` state which has implicit interactions. Makes following the logic very unpleasant. 

We can avoid the extra state by adding a parameter to `ins_complete`. @Shougo Can you try https://github.com/neovim/neovim/pull/4315 then we'll merge?

**Comment by Shougo** [comment]
@alexlafroscia @nhooyr @landaire Can you test #4315?

**Comment by nhooyr** [comment]
on it.

**Comment by alexlafroscia** [comment]
@Shougo that seems to be working for me, too

**Pull request closed without merging**